File: polyglot/ruby/quick_forensic_report.rb
# polyglot/ruby/quick_forensic_report.rb

require 'fileutils'
require 'digest'

class QuickForensicReport
  attr_reader :report_path, :signature_map, :results

  def initialize(report_path = 'forensic_report.txt')
    @report_path = report_path
    @signature_map = {
      'txt' => ['\x0D\x0A', '\x0A'],
      'jpg' => ['\xFF\xD8\xFF'],
      'png' => ['\x89\x50\x4E\x47\x0D\x0A\x1A\x0A'],
      'pdf' => ['\x25\x50\x44\x46'],
      'zip' => ['\x50\x4B\x03\x04'],
      'exe' => ['\x4D\x5A'],
      'elf' => ['\x7F\x45\x4C\x46']
    }
    @results = []
  end

  def generate_from_image(image_path)
    File.open(image_path, 'rb') do |file|
      buffer = ''
      while (chunk = file.read(1024))
        buffer += chunk
        process_buffer(buffer)
      end
    end
    write_report
  end

  private

  def process_buffer(buffer)
    @signature_map.each do |ext, signatures|
      signatures.each do |sig|
        if buffer.include?(sig)
          offset = buffer.index(sig)
          filename = "carved_file_#{Digest::SHA1.hexdigest(sig)}#{ext}"
          save_carved_file(buffer, offset, sig.size, filename)
          @results << { signature: sig.unpack('H*')[0], file: filename, offset: offset }
        end
      end
    end
  end

  def save_carved_file(data, offset, size, filename)
    FileUtils.mkdir_p(File.dirname(filename))
    File.open(filename, 'wb') do |f|
      f.write(data[offset...offset + size])
    end
  end

  def write_report
    File.open(@report_path, 'w') do |file|
      file.puts "Quick Forensic Report"
      file.puts "---------------------"
      file.puts "\nFound files:"
      @results.each do |result|
        file.puts "Signature: #{result[:signature]} (#{result[:signature].bytesize} bytes)"
        file.puts "File: #{result[:file]}"
        file.puts "Offset: #{result[:offset]}"
        file.puts "---------------------"
      end
      file.puts "\nReport saved to #{@report_path}"
    end
  end
end

# Entry point/demo
if __FILE__ == $0
  report = QuickForensicReport.new('forensic_report.txt')
  report.generate_from_image('disk_image.raw')
end

File: polyglot/cpp/quick_forensic_report.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <ctime>
#include <iomanip>

// Forward declaration
struct FileEntry;

// Function to read a file into a vector of bytes
std::vector<uint8_t> readFile(const std::string& filePath) {
    std::ifstream file(filePath, std::ios::binary | std::ios::ate);
    if (!file) {
        throw std::runtime_error("Failed to open file: " + filePath);
    }

    size_t fileSize = file.tellg();
    std::vector<uint8_t> buffer(fileSize);

    file.seekg(0, std::ios::beg);
    file.read(reinterpret_cast<char*>(buffer.data()), fileSize);

    return buffer;
}

// Function to find all occurrences of a signature in a byte buffer
std::vector<size_t> findSignatures(const std::vector<uint8_t>& buffer, const std::vector<uint8_t>& signature) {
    std::vector<size_t> positions;
    size_t sigLen = signature.size();
    size_t bufLen = buffer.size();

    for (size_t i = 0; i <= bufLen - sigLen; ++i) {
        bool match = true;
        for (size_t j = 0; j < sigLen; ++j) {
            if (buffer[i + j] != signature[j]) {
                match = false;
                break;
            }
        }
        if (match) {
            positions.push_back(i);
        }
    }

    return positions;
}

// Function to generate a quick forensic report
void generateQuickForensicReport(const std::vector<uint8_t>& buffer, const std::string& filePath) {
    std::cout << "=== Quick Forensic Report ===\n";
    std::cout << "Timestamp: " << std::put_time(std::localtime(&std::time(0)), "%Y-%m-%d %H:%M:%S") << "\n";
    std::cout << "File Path: " << filePath << "\n";
    std::cout << "Total Size: " << buffer.size() << " bytes\n";
    std::cout << "=== End of Header ===\n";

    // Example signatures for common file types
    std::vector<std::pair<std::string, std::vector<uint8_t>>> signatures = {
        {"JPEG", {0xFF, 0xD8, 0xFF, 0xE0, 0x00, 0x10, 0x4A, 0x46, 0x49, 0x46, 0x00, 0x01}},
        {"PNG", {0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A}},
        {"ZIP", {0x50, 0x4B, 0x03, 0x04}},
        {"PDF", {0x25, 0x50, 0x44, 0x46}},
        {"MP4", {0x00, 0x00, 0x00, 0x18, 0x66, 0x74, 0x79, 0x70, 0x33, 0x67, 0x72, 0x61, 0x70, 0x68, 0x32, 0x30}},
        {"TXT", {0x42, 0x4D}},
        {"EXE", {0x4D, 0x5A, 0x90}},
        {"ELF", {0x7F, 0x45, 0x4C, 0x46}},
        {"HTML", {0x3C, 0x21, 0x44, 0x4F, 0x43, 0x54, 0x3E}},
        {"CSV", {0x43, 0x57, 0x45, 0x42, 0x53, 0x59, 0x50, 0x41, 0x4E, 0x54}}
    };

    for (const auto& sig : signatures) {
        std::vector<size_t> positions = findSignatures(buffer, sig.second);
        if (!positions.empty()) {
            std::cout << "Found " << sig.first << " files:\n";
            for (size_t pos : positions) {
                std::cout << "  - Offset: " << pos << " bytes\n";
            }
        }
    }

    std::cout << "=== End of Report ===\n";
}

int main() {
    // Example usage with a sample file
    std::string filePath = "sample_disk_image.bin"; // Replace with actual disk image or memory dump path

    try {
        std::vector<uint8_t> buffer = readFile(filePath);
        generateQuickForensicReport(buffer, filePath);
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }

    return 0;
}